Added wrappers to some parameters to the camera.

GuillaumeSchmid opened the pull request:
* Data move policy: buffer_policy
* Auto white balance mode: auto_wb
* White balance Red coefficient: wb_kr
* White balance Green coefficient: wb_kg
* White balance Blue coefficient: wb_kb

GuillaumeSchmid committed:
Added the following parmeters to the camera:

* Data move policy: buffer_policy
* Auto white balance mode: auto_wb
* White balance Red coefficient: wb_kr
* White balance Green coefficient: wb_kg
* White balance Blue coefficient: wb_kb

GuillaumeSchmid committed:
Added exposure and gain setting in direct mode.

GuillaumeSchmid committed:
Removed dead code.

GuillaumeSchmid committed:
Refactored code to use a single function to append param  modifers before call to setParam.

GuillaumeSchmid committed:
Refactored code to use a single function to append param  modifers before call to setParam.

GuillaumeSchmid committed:
Small fix on param name, forgot \0

GuillaumeSchmid committed:
Small fix on param name, forgot \0

GuillaumeSchmid committed:
Try tu understand th panic on converting b"exposure: direct_update\0"
to cstr.

superbort commented:
Hi,
most of those changes look good to me. I was going to merge them (after testing), but now the request has been closed. Is there anything I can do to help?

GuillaumeSchmid commented:
Hello!
I am thrilled that you are considering my PR. That is cool, as I need those
features for our system.
I implemented more features in the library and did not split them into
multiple PR. I also have a weird bug I tried to fix in set_gain that
somehow made the PR more complex than what I would like.

I have 2 problems. You may have some clues on how to fix them.
1 - On the Acquisition Buffer, I implemented the set_gain and set_exposure.
The problem is that the function works if I do this:
 xiapi_sys::xiSetParamInt(self.camera.device_handle, b"exposure:
direct_update\0".as_ptr().cast(), exposure as i32) };
The weird thing is the space between exposure: and direct_update. If I use
the regular method (param_info) to merge XI_PRM_EXPOSURE and
XI_PRMM_DIRECT_UPDATE, it does not work
I have an unknown parameter error. But with this code, it works.

2 - On the set_gain function if I do this:
// param suffix is param_info renamed to be more generic.
let param_name = unsafe{ param_suffix(XI_PRM_GAIN,
XI_PRMM_DIRECT_UPDATE).unwrap()};
        let param_c = CStr::from_bytes_with_nul( param_name.as_bytes()
).unwrap();
        let err = unsafe {
xiapi_sys::xiSetParamFloat(self.camera.device_handle, param_c.as_ptr(),
gain) };

I have an invalid parameter error. So the param_c is not the problem, but I
don't understand what is happening.

Also, not related to this problem, sometimes, I have another error:
xiAPI: SAL_AR1820HS_CalculateAndSetSensorPLL line_lenght_pck is too short.
When I start the acquisition: no frames.
I did not find anything about this error in the documentation. Do you know
what is happening ?

I will redo the PR and try to split the features properly. I will do it
quickly and republish.

Yours,
Guillaume.

Le lun. 6 nov. 2023 à 14:04, Alexander Brömmer ***@***.***> a
écrit :

> Hi,
> most of those changes look good to me. I was going to merge them (after
> testing), but now the request has been closed. Is there anything I can do
> to help?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Ximea-GmbH/xiapi/pull/2#issuecomment-1794770767>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEM32SJ7EKJ6YO7NSL3TRH3YDDN4PAVCNFSM6AAAAAA6VYFOMGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTOOJUG43TANZWG4>
> .
> You are receiving this because you modified the open/close state.Message
> ID: ***@***.***>
>

GuillaumeSchmid commented:
Sorry for the follow-up, I forgot to mention that I also implemented in my
code a function that converts error code to string messages.
I may have missed it, but I did not see that in the sys API.
I could also make a PR if you want to integrate this in the library.

Yours,
 Guillaume.


Le lun. 6 nov. 2023 à 14:22, Guillaume Schmid ***@***.***> a
écrit :

> Hello!
> I am thrilled that you are considering my PR. That is cool, as I need
> those features for our system.
> I implemented more features in the library and did not split them into
> multiple PR. I also have a weird bug I tried to fix in set_gain that
> somehow made the PR more complex than what I would like.
>
> I have 2 problems. You may have some clues on how to fix them.
> 1 - On the Acquisition Buffer, I implemented the set_gain and set_exposure.
> The problem is that the function works if I do this:
>  xiapi_sys::xiSetParamInt(self.camera.device_handle, b"exposure:
> direct_update\0".as_ptr().cast(), exposure as i32) };
> The weird thing is the space between exposure: and direct_update. If I use
> the regular method (param_info) to merge XI_PRM_EXPOSURE and
> XI_PRMM_DIRECT_UPDATE, it does not work
> I have an unknown parameter error. But with this code, it works.
>
> 2 - On the set_gain function if I do this:
> // param suffix is param_info renamed to be more generic.
> let param_name = unsafe{ param_suffix(XI_PRM_GAIN,
> XI_PRMM_DIRECT_UPDATE).unwrap()};
>         let param_c = CStr::from_bytes_with_nul( param_name.as_bytes()
> ).unwrap();
>         let err = unsafe {
> xiapi_sys::xiSetParamFloat(self.camera.device_handle, param_c.as_ptr(),
> gain) };
>
> I have an invalid parameter error. So the param_c is not the problem, but
> I don't understand what is happening.
>
> Also, not related to this problem, sometimes, I have another error:
> xiAPI: SAL_AR1820HS_CalculateAndSetSensorPLL line_lenght_pck is too short.
> When I start the acquisition: no frames.
> I did not find anything about this error in the documentation. Do you know
> what is happening ?
>
> I will redo the PR and try to split the features properly. I will do it
> quickly and republish.
>
> Yours,
> Guillaume.
>
> Le lun. 6 nov. 2023 à 14:04, Alexander Brömmer ***@***.***>
> a écrit :
>
>> Hi,
>> most of those changes look good to me. I was going to merge them (after
>> testing), but now the request has been closed. Is there anything I can do
>> to help?
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/Ximea-GmbH/xiapi/pull/2#issuecomment-1794770767>, or
>> unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AEM32SJ7EKJ6YO7NSL3TRH3YDDN4PAVCNFSM6AAAAAA6VYFOMGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTOOJUG43TANZWG4>
>> .
>> You are receiving this because you modified the open/close state.Message
>> ID: ***@***.***>
>>
>
